File: scripts/tools/validate_module_headers.py
#!/usr/bin/env python3
"""Validate Surge/Shadowrocket module header integrity.

Guards against iOS client "empty module fields" regressions caused by malformed
metadata blocks (for example, stray non-comment lines before the first section).
"""

from __future__ import annotations

from pathlib import Path
import re
import sys

ROOT = Path(__file__).resolve().parents[2]
MODULE_ROOTS = [
    ROOT / "module" / "surge(main)",
    ROOT / "module" / "shadowrocket",
]

META_RE = re.compile(r"^#!\s*([A-Za-z0-9_-]+)\s*=\s*.*$")
RECOMMENDED_KEYS = ("name", "desc", "author", "category")
MAX_META_LINE_LENGTH = 4096


def validate_module(path: Path) -> tuple[list[str], list[str]]:
    issues: list[str] = []
    warnings: list[str] = []
    try:
        lines = path.read_text(encoding="utf-8").splitlines()
    except Exception as exc:  # pragma: no cover - best effort diagnostics
        return [f"{path}: failed to read file: {exc}"], []

    if not lines:
        return [f"{path}: empty module file"], []

    in_header = True
    seen_section = False
    seen_keys: set[str] = set()

    for lineno, raw in enumerate(lines, 1):
        stripped = raw.lstrip("\ufeff").strip()

        if in_header and stripped.startswith("[") and stripped.endswith("]"):
            in_header = False
            seen_section = True
            continue

        if in_header:
            if not stripped:
                continue
            if stripped.startswith("#!"):
                if len(stripped) > MAX_META_LINE_LENGTH:
                    issues.append(f"{path}:{lineno}: metadata line too long (> {MAX_META_LINE_LENGTH})")
                m = META_RE.match(stripped)
                if not m:
                    issues.append(f"{path}:{lineno}: malformed metadata line")
                    continue
                seen_keys.add(m.group(1).lower())
                continue
            if stripped.startswith("#"):
                continue
            issues.append(
                f"{path}:{lineno}: illegal header content before first section: {stripped[:80]}"
            )
            continue

    if not seen_section:
        issues.append(f"{path}: missing any [Section] block")

    missing = [k for k in RECOMMENDED_KEYS if k not in seen_keys]
    if missing:
        warnings.append(f"{path}: missing recommended metadata keys: {', '.join(missing)}")

    return issues, warnings


def main() -> int:
    all_issues: list[str] = []
    all_warnings: list[str] = []
    for base in MODULE_ROOTS:
        if not base.exists():
            all_issues.append(f"{base}: directory not found")
            continue
        pattern = "*.sgmodule" if "surge(main)" in str(base) else "*.module"
        for fp in sorted(base.rglob(pattern)):
            issues, warnings = validate_module(fp)
            all_issues.extend(issues)
            all_warnings.extend(warnings)

    if all_issues:
        print("❌ Module header validation failed:")
        for issue in all_issues:
            print(f"  - {issue}")
        return 1

    if all_warnings:
        print("⚠️ Module header validation warnings:")
        for warning in all_warnings:
            print(f"  - {warning}")

    print("✅ Module header validation passed.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: create/qa/validate_module_headers.go
package qa

import (
	"fmt"
	"io/fs"
	"path/filepath"
	"regexp"
	"strings"
	"unicode/utf8"

	"github.com/nyamiiko/script_hub/go_scripts/hub"
)

var metaRe = regexp.MustCompile(`^#!\s*([A-Za-z0-9_-]+)\s*=.*$`)

func validateModule(path string) ([]string, []string) {
	var issues []string
	var warnings []string

	text := hub.ReadFileString(path)
	if text == "" {
		return []string{fmt.Sprintf("%s: empty module file", path)}, nil
	}

	lines := strings.Split(text, "\n")
	inHeader := true
	seenSection := false
	seenKeys := make(map[string]bool)

	for lineno, raw := range lines {
		raw = strings.TrimPrefix(raw, "\ufeff")
		stripped := strings.TrimSpace(raw)

		if inHeader && strings.HasPrefix(stripped, "[") && strings.HasSuffix(stripped, "]") {
			inHeader = false
			seenSection = true
			continue
		}

		if inHeader {
			if stripped == "" {
				continue
			}
			if strings.HasPrefix(stripped, "#!") {
				if utf8.RuneCountInString(stripped) > 4096 {
					issues = append(issues, fmt.Sprintf("%s:%d: metadata line too long (> 4096)", path, lineno+1))
				}
				m := metaRe.FindStringSubmatch(stripped)
				if m == nil {
					issues = append(issues, fmt.Sprintf("%s:%d: malformed metadata line", path, lineno+1))
					continue
				}
				seenKeys[strings.ToLower(m[1])] = true
				continue
			}
			// Treat plain text in header as implicit description/comments
			continue
		}
	}

	if !seenSection {
		issues = append(issues, fmt.Sprintf("%s: missing any [Section] block", path))
	}

	recommendedKeys := []string{"name", "desc", "author", "category"}
	var missing []string
	for _, k := range recommendedKeys {
		if !seenKeys[k] {
			missing = append(missing, k)
		}
	}
	if len(missing) > 0 {
		warnings = append(warnings, fmt.Sprintf("%s: missing recommended metadata keys: %s", path, strings.Join(missing, ", ")))
	}

	return issues, warnings
}

func RunValidateModuleHeaders() int {
	moduleRoots := []string{
		filepath.Join(hub.ROOT, "modules/surge"),
		filepath.Join(hub.ROOT, "modules/shadowrocket"),
	}

	var allIssues []string
	var allWarnings []string

	for _, base := range moduleRoots {
		if !hub.ValidateDirExists(base, "") {
			allIssues = append(allIssues, fmt.Sprintf("%s: directory not found", base))
			continue
		}
		pattern := "*.module"
		if strings.Contains(filepath.ToSlash(base), "modules/surge") {
			pattern = "*.sgmodule"
		}

		filepath.WalkDir(base, func(path string, d fs.DirEntry, err error) error {
			if err != nil {
				return nil
			}
			matched, _ := filepath.Match(pattern, d.Name())
			if !d.IsDir() && matched {
				issues, warnings := validateModule(path)
				allIssues = append(allIssues, issues...)
				allWarnings = append(allWarnings, warnings...)
			}
			return nil
		})
	}

	if len(allIssues) > 0 {
		fmt.Println("❌ Module header validation failed:")
		for _, issue := range allIssues {
			fmt.Printf("  - %s\n", issue)
		}
		return 1
	}

	if len(allWarnings) > 0 {
		fmt.Println("⚠️ Module header validation warnings:")
		for _, warning := range allWarnings {
			fmt.Printf("  - %s\n", warning)
		}
	}

	fmt.Println("✅ Module header validation passed.")
	return 0
}
